src/routing/routes.ts

```diff
@@ -0,0 +1,114 @@
+import { HttpMethod, HttpRequest, HttpResponse } from "../http";
+import { HttpHandler } from "../http4ts";
+import { pathToRegexp, Key } from "./path-to-regexp";
+
+export interface RoutedHttpRequest extends HttpRequest {
+  routeParams: Record<string, string>;
+}
+
+export type RoutedHttpHandler = (
+  req: RoutedHttpRequest
+) => HttpResponse | Promise<HttpResponse>;
+
+export type UriTemplate = string;
+
+interface RouteDefinition {
+  method?: HttpMethod;
+  uriTemplate?: RegExp;
+  keys?: Key[];
+  handler: RoutedHttpHandler;
+}
+
+function defaultNotFoundHandler(): HttpResponse {
+  return {
+    body: "Not Found",
+    headers: {},
+    status: 404
+  };
+}
+export function routes(...routeDefinitions: RouteDefinition[]): HttpHandler {
+  let notFoundHandler: RoutedHttpHandler = defaultNotFoundHandler;
+  const lastDefinition = routeDefinitions[routeDefinitions.length - 1];
+  if (
+    !lastDefinition.method &&
+    !lastDefinition.keys &&
+    !lastDefinition.uriTemplate
+  ) {
+    notFoundHandler = lastDefinition.handler;
+  }
+
+  return req => {
+    const foundRoute = routeDefinitions.find(rd => {
+      if (rd.method && rd.uriTemplate) {
+        return rd.method === req.method && rd.uriTemplate.test(req.url);
+      } else if (rd.uriTemplate) {
+        return rd.uriTemplate.test(req.url);
+      }
+    });
+
+    if (foundRoute) {
+      if (!foundRoute.uriTemplate || !foundRoute.keys) {
+        throw new Error("Route should be templated.");
+      }
+      const result = foundRoute.uriTemplate.exec(req.url);
+      if (!result) {
+        throw new Error("Route didn't match!");
+      }
+      const routeParams = foundRoute.keys.reduce(
+        (prev, cur, i) => ({ ...prev, [cur.name]: result[i + 1] }),
+        {}
+      );
+      return foundRoute.handler({ ...req, routeParams });
+    } else {
+      return notFoundHandler({ ...req, routeParams: {} });
+    }
+  };
+}
+
+export function def(
```

Comment:
If we are exposing def, than maybe we can decide on a better name

Reply:
How about one of these? `defRoute`, `defineRoute`, `use`

Reply:
What about just: `Route`